src/lightning_lite/accelerators/cuda.py

```diff
@@ -77,6 +78,25 @@ def _get_all_available_cuda_gpus() -> List[int]:
     return list(range(num_cuda_devices()))
 
 
+@contextmanager
+def _patch_cuda_is_available() -> Generator:
+    """Context manager that safely patches :func:`torch.cuda.is_available` with its NVML-based version if
+    possible."""
+    orig_check = None
+    new_check = torch.cuda.device_count if _TORCH_GREATER_EQUAL_1_13 else _device_count_nvml
+
+    if hasattr(torch._C, "_cuda_getDeviceCount") and _device_count_nvml() >= 0:
+        # we can safely patch is_available if both torch has CUDA compiled and the NVML count is succeeding
+        # otherwise, patching is_available could lead to attribute errors or infinite recursion
+        orig_check = torch.cuda.is_available
+        torch.cuda.is_available = new_check  # type: ignore[assignment]
```

Comment:
I'm a bit confused. Why do we want to assign `device_count()` to `is_available`? Wouldn't we need to convert to bool?

Reply:
Since the returned `int` would never be negative in this context, I was avoiding the unnecessary check using the truth value of the returned int. Using `is_cuda_available` may be cleaner though so using that with the latest commit (after rebasing).